### PR TITLE
Anti affinity selectors for pod

### PR DESCRIPTION
**Pull request by ravigadde**
Carried over from #9560

cc @bgrant0607 @davidopp @sdminonne

### CODE REVIEW

**Comment by ravigadde** [comment]
Should LabelSelector move from labels pkg to api and become a versioned object?

File: pkg/api/v1/types.go [hunk 1]
```diff
@@ -1205,6 +1206,8 @@ type PodSpec struct {
 	// Selector which must match a node's labels for the pod to be scheduled on that node.
 	// More info: http://releases.k8s.io/HEAD/docs/user-guide/node-selection/README.md
 	NodeSelector map[string]string `json:"nodeSelector,omitempty"`
+	// AntiAffinitySelectors are selectors which must not match against existing pods labels for the pod to fit on a node
+	AntiAffinitySelectors []labels.LabelSelector `json:"antiAffinitySelectors,omitempty"`
```
Comment (bgrant0607) [inline pkg/api/v1/types.go]: Did you deliberately drop the change to PodConflictSelectors?

**Comment by bgrant0607** [comment]
Yes, we need a serializable, properly documented, properly tagged type definition of LabelSelector in the v1 API.

File: pkg/api/v1/types.go (hunk 1 above)
Comment (ravigadde) [inline pkg/api/v1/types.go]: Oops.. apologies, the name changed so many times, I lost track.

**Comment by ravigadde** [comment]
Ok, let me make those changes and update the PR

**Comment by thockin** [comment]
Surely there is a doc somewhere that you can write about this feature.  Also maybe an example?

**Comment by HaiyangDING** [comment]
CC @HaiyangDING

**Comment by bgrant0607** [comment]
We're currently focused on 1.1. 

Also, I think we need to take a step back and figure out a rough idea of what scheduling policies we're going to want and how/where they should be expressed (#14816), and how we're going to evolve those policies without breaking every third-party scheduler in existence. 

Furthermore, there are big changes we want to make to the scheduler in order to make it easier to add more factors to the objective function and to improve performance/scalability. I'm concerned that adding more features now will make these changes harder.

Finally, we need to start with just the selector API addition in another PR. We need that for all our controllers and to extend the current nodeSelector. We can't just reuse the type from pkg/labels. That type isn't properly tagged and documented, and won't have the proper conversion, deepcopy, swagger, etc. generated.

**Comment by bgrant0607** [comment]
We're going to work on #341.

**Comment by mynameismevin** [comment]
Hey guys. thanks for the effort put into this issue thus far. as there are a lot of changes that are coming post 1.0, I was curious to see what the current planned release would be for this functionality? I saw that you have focus on the 1.1 scheduler changes, does that include these anti-affinity features?

**Comment by davidopp** [comment]
I assume you mean "1.1" where you wrote "1.0" and "1.2" where you wrote "1.1"

We may have this feature in 1.2 if we can shepherd the community contributions through in that timeframe. I think we would want to do NodeSelector at the same time.
https://github.com/kubernetes/kubernetes/issues/341#issuecomment-154663349

**Comment by davidopp** [comment]
Closing because this is subsumed by #18265. Please comment on that PR if you think there is a use case you need that it does not cover.

**Pull request closed without merging**